Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +31,142 @@ public class DafnyHoverHandler : HoverHandlerBase {
       };
     }
 
-    public async override Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
+    public override async Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
       logger.LogTrace("received hover request for {Document}", request.TextDocument);
       var document = await documents.GetDocumentAsync(request.TextDocument);
       if (document == null) {
         logger.LogWarning("the document {Document} is not loaded", request.TextDocument);
         return null;
       }
+      var diagnosticHoverContent = GetDiagnosticsHover(document, request.Position, out var areMethodStatistics);
       if (!document.SymbolTable.TryGetSymbolAt(request.Position, out var symbol)) {
         logger.LogDebug("no symbol was found at {Position} in {Document}", request.Position, request.TextDocument);
+      }
+
+      var symbolHoverContent = symbol != null ? CreateSymbolMarkdown(symbol, cancellationToken) : null;
+      if (diagnosticHoverContent == null && symbolHoverContent == null) {
         return null;
       }
-      return CreateHover(symbol, cancellationToken);
+
+      // If diagnostics are method diagnostics, we prioritize displaying the symbol information.
+      // This makes testing easier and less surprise for the user.
+      var hoverContent = areMethodStatistics && symbolHoverContent != null ? "" : (diagnosticHoverContent ?? "");
+      hoverContent = symbolHoverContent != null ? hoverContent + (hoverContent != "" ? "  \n" : "") + symbolHoverContent : hoverContent;
+      return CreateMarkdownHover(hoverContent);
+    }
+
+    private string? GetDiagnosticsHover(DafnyDocument document, Position position, out bool areMethodStatistics) {
+      areMethodStatistics = false;
+      foreach (var node in document.VerificationTree.Children.OfType<TopLevelDeclMemberVerificationTree>()) {
```

Comment:
Eventually it might be possible to use a more efficient data structure for this, but that's probably only worth it if we find that this repeated iteration is too expensive.

Reply:
In the past I've done this type of thing by first finding a particular AST element, in this case it would be finding the assertion statement, which you do by checks of the `node.Range.Contains(position)` starting at the root node of the tree. That's similar to what's being done here although it would include modules and classes as filters. I think for finding elements in lists, like finding a particular statement in a list of statements, you can also do a binary search instead of a linear traversal.

Then given a particular AST node, compilation results relating to that AST node, such as resolved declarations for nodes that are references, or in this case verification performance data, should be attached to it using some form of a dictionary.

@MikaelMayer have you considered storing the verification results on the AST instead of on a separate tree that mimics parts of the AST?

Reply:
I haven't considered changing the AST to store verification results there, not just because I think it is better if we modified the AST as little as possible,
but especially because the AST does not have the notion of implementations and assertion batches like what we get from verification.

Reply:
> especially because the AST does not have the notion of implementations and assertion batches like what we get from verification.

Good point. Food for thought :-)